drivers/apds99xx/apds99xx.c

```diff
@@ -0,0 +1,526 @@
+/*
+ * Copyright (C) 2018 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_apds99xx
+ * @brief       Device driver for the Broadcom APDS99XX proximity and ambient light sensor
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ * @file
+ * @{
+ */
+
+#include <string.h>
+#include <stdlib.h>
+
+#include "apds99xx_regs.h"
+#include "apds99xx.h"
+
+#include "irq.h"
+#include "log.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+
+#define DEBUG_DEV(f, d, ...) \
+        DEBUG("[apds99xx] %s i2c dev=%d addr=%02x: " f "\n", \
+              __func__, d->params.dev, APDS99XX_I2C_ADDRESS, ## __VA_ARGS__);
+
+#else /* ENABLE_DEBUG */
+
+#define DEBUG_DEV(f, d, ...)
+
+#endif /* ENABLE_DEBUG */
+
+#define ERROR_DEV(f, d, ...) \
+        LOG_ERROR("[apds99xx] %s i2c dev=%d addr=%02x: " f "\n", \
+                  __func__, d->params.dev, APDS99XX_I2C_ADDRESS, ## __VA_ARGS__);
+
+/** Forward declaration of functions for internal use */
+
+static int _is_available(const apds99xx_t *dev);
+static void _set_reg_bit(uint8_t *byte, uint8_t mask, uint8_t bit);
+static int _reg_read(const apds99xx_t *dev, uint8_t reg, uint8_t *data, uint16_t len);
+static int _reg_write(const apds99xx_t *dev, uint8_t reg, uint8_t *data, uint16_t len);
+static int _update_reg(const apds99xx_t *dev, uint8_t reg, uint8_t mask, uint8_t val);
+
+int apds99xx_init(apds99xx_t *dev, const apds99xx_params_t *params)
+{
+    /* some parameter sanity checks */
+    assert(dev != NULL);
+    assert(params != NULL);
+    assert(params->als_steps <= 256);
+    assert(params->wait_steps <= 256);
+#if MODULE_APDS9960
+    assert(params->prx_pulses <= 15);
+#endif
+
+    DEBUG_DEV("params=%p", dev, params);
+
+    /* init sensor data structure */
+    dev->params = *params;
+
+#if MODULE_APDS99XX_FULL
+    dev->isr = NULL;
+    dev->isr_arg = NULL;
+    dev->gpio_init = false;
+#endif
+
+    /* wait for 6 ms after power on reset */
+    xtimer_usleep(6 * US_PER_MS);
```

Comment:
would be nice to get rid of `xtimer`  usage, especially as this is the only one. Isn't there a state register or gpio that we could poll (`while(...) {}` style) instead?

Reply:
The problem is that this time is specified in the data sheet. It is the time the sensor needs for initialization after power-on to become ready for operation. Unfortunately there is no status flag.

Possible solutions:

1. We simply assume that RIOT takes longer to load than this 6 ms before the device is accessed for the first time.

2. We allow multiple tries of the `_is_available' function and use the  I2C ACK to determine when the device is available. However, at an I2C speed of 400 kbps, each failed attempt on the bus would only take 0.025 ms, which could mean a maximum of 250 failed attempts for 6 ms.

Reply:
I have removed the dependency on xtimer with commit https://github.com/RIOT-OS/RIOT/pull/10420/commits/d909fad27cd9d2665c11d2b464be2d8df8005be2.